[wip] server: add AllocateNodeID endpoint

tbg opened the pull request:
Currently, when a node first joins an existing cluster, it goes through
a very awkward dance in which it first runs without a NodeID until it
is nominally connected to the cluster, then requests a NodeID via its
semi-initialized DB and then goes and fully initializes itself.

This makes the NodeID late-bound and since we use the NodeID in lots of
places, lots of places look really awkward and have to juggle the fact
that the NodeID may not yet be set and is expected to change later.

This PR doesn't change anything about that, BUT it adds a gRPC endpoint
that you hit and which gives you a NodeID back. This endpoint is unused
at the moment but by shipping it in release X, we may assume in release
X+1 that all nodes we interact with do support it. In particular,
release X+1 is free to rework the initialization sequence so that it
looks as follows:

1. decide that joining an existing cluster is necessary
2. grab the `--join` flags
3. hit the AllocateNodeID endpoint for the first target using
   a bare bones gRPC connection (i.e. don't go through *rpc.Context)
   using a fairly generous timeout
4. on success, done - on error, iterate through the other join flags,
   wrapping around and logging warnings if necessary
5. once a NodeID has been allocated, go ahead and start bootstrapping
   the local state.

This is much simpler and lots of places in our code will be better off
for it.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36921)
<!-- Reviewable:end -->

tbg committed:
server: add AllocateNodeID endpoint

Currently, when a node first joins an existing cluster, it goes through
a very awkward dance in which it first runs without a NodeID until it
is nominally connected to the cluster, then requests a NodeID via its
semi-initialized DB and then goes and fully initializes itself.

This makes the NodeID late-bound and since we use the NodeID in lots of
places, lots of places look really awkward and have to juggle the fact
that the NodeID may not yet be set and is expected to change later.

This PR doesn't change anything about that, BUT it adds a gRPC endpoint
that you hit and which gives you a NodeID back. This endpoint is unused
at the moment but by shipping it in release X, we may assume in release
X+1 that all nodes we interact with do support it. In particular,
release X+1 is free to rework the initialization sequence so that it
looks as follows:

1. decide that joining an existing cluster is necessary
2. grab the `--join` flags
3. hit the AllocateNodeID endpoint for the first target using
   a bare bones gRPC connection (i.e. don't go through *rpc.Context)
   using a fairly generous timeout
4. on success, done - on error, iterate through the other join flags,
   wrapping around and logging warnings if necessary
5. once a NodeID has been allocated, go ahead and start bootstrapping
   the local state.

This is much simpler and lots of places in our code will be better off
for it.

Release note: None

tbg commented:
Ben reminded me of https://github.com/cockroachdb/cockroach/pull/35367#issuecomment-470076965, which I feel silly about forgetting, but also I really regret my opinion there. We should've merged (some variation of) that PR.

tbg commented:
@bdarnell points out that we may get away with not allowing X+1 nodes to join an X/X+1 mixed cluster. We should figure that out because if we don't need to support that, we can include the steps in the commit message in the X release already instead of waiting to X+1. Either way, we should prototype the switchover to see if there are any unexpected problems with the approach.

BTW, `X` here is 19.2 (nothing will go into 19.1 at this point, @bdarnell tells me).

tbg commented:
(marking as WIP to reflect the experimental status)
